### PR TITLE
Implement invulnerability after shrinking

### LINKED ISSUE #535: [Game Play] Too much 'grow' and 'shrink'
Players become very large when it collides with other players. The same is true for the opposite, as the other player shrinks greatly.
We can adjust the effects of 'grow' and 'shrink', to a reasonable level

### LINKED ISSUE #699: Expand `player_state`
Currently, `player_state` has the following members:
```c++
enum class player_state
{
  active,
  dead
};
```
My proposed change:
```c++
enum class player_state
{
  active, // a normal state
  passive, // a state that the player is non-interactable
  dead // deactivated
};
```
A passive state is that a player is invulnerable **_and_** cannot interact with the other players/enviroment, we should distinguish this from the real invulnerable state. When a player shrinks, it should become passive, not only invulnerable.

### PR DESCRIPTION
**Pull request by EvoLandEco**
fixes #535 
The logic is:

1. When a player loses, it shrinks and becomes passive for a while
2. When a player is passive, it cannot grow or shrink, it cannot make other players grow or shrink either, it will be like a ghost
3. After a while, an passive player will become active again

### CODE REVIEW

**Comment by TheoPannetier** [comment]
I've assigned you to #381 #382 #463; please make sure that you assign yourself to issues before you start working on them! Otherwise other people may decide to work on it too, come up with a different solution, eventually causing a merge conflict (and some wasted time)!

**Comment by EvoLandEco** [comment]
> I've assigned you to #381 #382 #463; please make sure that you assign yourself to issues before you start working on them! Otherwise other people may decide to work on it too, come up with a different solution, eventually causing a merge conflict (and some wasted time)!

Yes, I should have done it before this PR. I also assigned myself #469 😸

**Comment by EvoLandEco** [comment]
When I was thinking about adding visual effect for invulnerability, I feel like the current implementation is bad. The reason is, we already have a `player_state` class, a better option would be to expand this enum class, add some states into it, and build the whole thing on top of it, rather than inventing a new flag `m_invulnerability` in the `player` class.

@TheoPannetier @swom apologies for inviting you as reviewers prematurely and (maybe) have already wasted a bit of your time. I will re-implement invulnerability,

File: src/player.cpp
Comment (EvoLandEco) [inline src/player.cpp:133]: Things become more complicated here: how do we define colliding? Simply two players are overlapping? Or they should all be non-passive and/or non-invulnerable? Here I chose to define colliding as they are simply overlapping, so as long as they are not out/dead, they can collide. This will not break the other part of the code. We might need to discuss those mechanics in depth later. 

We could add a feature later to make non-passive players unable to overlap with each other,

Comment (TheoPannetier) [reply src/player.cpp:133]: Yeah I think it confused me too the first time I saw this; colliding means overlapping AND active, i.e. capable to interact. If two players (or other objects) are overlapping but one or either is not active, there is no collision. I can't think of a better name for this, unfortunately!

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:671]: I don't understand why you changed the tests here. I see that you're being more exhaustive, systematically asserting that all setters exist. I don't see the necessity for this; the short test I had written was enough to reach the goal of the issue ("A player can become passive/invulnerable"). Being able to turn it back is not a necessity to reach this.
I believe (I'm not sure, can't read it from your commit history) what you are doing here is characteristic of unit-testing, that is writing tests, typically after the code, to ensure that the concerned bit of code keeps working as intended. This is ok, but we prefer test-driven development. The [difference ](https://methodpoet.com/unit-testing-vs-tdd/) is that in TDD, we write the test *before* the code and even before thinking of the implementation. The test states our intended goal, and will pass when this goal is reached. Then we seek to implement a solution.

This is good work, your tests are functional and do satisfy the end goal, I just want to raise the point that I believe this is excessive, maybe premature testing.

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:675]:
```suggestion
  /// A passive player cannot grow
```

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:699]:
```suggestion
  /// A passive player cannot shrink
```

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:727]:
```suggestion
  // Can get and set a player's passive state duration
```

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:738]: I can't see why you would change the duration of the passive state at the player level; shouldn't it be a fixed parameter of the game?

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:756]: Same comment as above for test 381; I don't think this adds anything that is not already (implicitly) required by the test I had written (below)

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:805]: Same as above, I don't think it's necessary to explicitly test for the timer?
Unless you think the issue needed to be broken down in smaller steps - in which case, sure, go ahead and create smaller intermediate issues towards losing invulnerability after some time :)

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:341]: I think passive states shouldn't be caught here; instead if either player is passive, there is no collision (and thus, growing/shrinking won't be called)

File: src/game.cpp
Comment (TheoPannetier) [inline src/game.cpp:147]: I disagree with the refactoring of `invulnerable` into `passive`. I think the former conveys more explicit information than the latter. `passive` is too general, could refer to the player being inactive for any reason, e.g. having been killed, stunned, or recovering from a defeat, and suggest the player cannot interact with anything, including foods etc. `invulnerable` simply means the player cannot be attacked, and lose a fight

**Review by TheoPannetier — CHANGES_REQUESTED** [review]
I have been a bit harsh and grumpy in my comments, but overall this is good work and this contribution definitely helps the game move forward :)
So there is only one point where I request a change before merging: collision should not happen if either member is passive, and in turn there is no need to check for state in grow and shrink.
I disagree with the name passive instead of invulnerable, but the last word is yours for this :+1:

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:671]: @TheoPannetier Thanks for pointing out this, it was that I feel like the tests I had written beforehand were not enough to detect all the changes in the future that might be done intentionally or unintentionally by me or by others, I am sometimes too obsessive that I force myself to add a lot of probably excessive or premature tests. 😢

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:738]: @TheoPannetier I was thinking about possible future extension of the power-ups etc. For a minimal viable game, it should also be good to just let the duration be a fixed parameter of the game. What is your opinion about this?

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:805]: I added these tests for debugging purposes as I encountered quite a lot of issues. I was making sure that all the parts I had added were working as I expected.

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:147]: IMO we do need to have some sound definitions of these states. For the `passive` state, I was referring to a game state that is the same as in GTA5, which means a player is alive but cannot be killed by the others players, and of course, a passive player cannot hurt the other players. A passive player can collect things and do jobs.

Considering foods or power-ups, it is worthwhile to think thoroughly. Maybe we could find a more proper name for this state I was referring to? As for `invulnerable`, I still feel like it could be a state after collecting a power-up or triggering some game mechanics.

**Comment by EvoLandEco** [comment]
@TheoPannetier An open discussion is always helpful, thanks for the opinions, and let's keep talking and find a best solution 😉

File: src/game.cpp
Comment (TheoPannetier) [reply src/game.cpp:671]: >  I am sometimes too obsessive

I definitely can't blame you for *caring* about the future of the game! I confirm I was too harsh with my initial comment, and I'd like to apologise for this. 

These tests are indeed a bit premature at this stage, but they don't remove anything or add anything dangerous to the game AFAICS and may pave the way for the next steps with the passive state. Take the above as advice on how to better design tests, but what you're doing is an important step too.

File: src/game.cpp
Comment (TheoPannetier) [reply src/game.cpp:738]: I think you've figured out my opinion already 😉 
I believe anticipating possible extensions beyond the minimal viable game is premature optimisation, and a wise man once told me it's the [root of all evil](https://effectiviology.com/premature-optimization/).

So for now, I'd rather have the duration as a fixed parameter. It will be easy to make it free again when/if we need it :)

File: src/game.cpp
Comment (TheoPannetier) [reply src/game.cpp:805]: The main test (if it's good!) should already ensure that everything works as expected. If players lose invulnerability after some time and not earlier, as #463 demands, then we have our feature and we're happy!
I'd argue we only care about the underlying mechanisms and potential issues *if* a problem indeed manifests after completing the test. Then it's time to make a new issue to signal the problem, and write a test that ensures it is fixed.

That being said, using tests as debug towards completing another test is entirely fine, but then no need to leave it in the code once you have resolved the final test!

File: src/game.cpp
Comment (TheoPannetier) [reply src/game.cpp:147]: Fair enough, I agree with the reasoning, although again I prefer `invulnerable` over `passive` to refer to what you describe: `passive` implies you cannot do anything other than move, so no item or task collection. On the other hand, `invulnerable` does suggest you could still attack other players, which is not true (or is it?). 
I'll let you make the final call :)

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:805]: > The main test (if it's good!) should already ensure that everything works as expected. If players lose invulnerability after some time and not earlier, as #463 demands, then we have our feature and we're happy! I'd argue we only care about the underlying mechanisms and potential issues _if_ a problem indeed manifests after completing the test. Then it's time to make a new issue to signal the problem, and write a test that ensures it is fixed.
> 
> That being said, using tests as debug towards completing another test is entirely fine, but then no need to leave it in the code once you have resolved the final test!

Yes that's true, I mean I would add extra debugging tests besides the main tests and remove them right before merging, this could be a better way than what was done in this PR 😉

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:147]: > Fair enough, I agree with the reasoning, although again I prefer `invulnerable` over `passive` to refer to what you describe: `passive` implies you cannot do anything other than move, so no item or task collection. On the other hand, `invulnerable` does suggest you could still attack other players, which is not true (or is it?). I'll let you make the final call :)

I think what is important is to have two types of invulnerable (not receiving damage) states, one you can still deal damage while the other one you cannot.

To be more specific:

- Invulnerable State A: If a player receives damage, it shrinks and becomes State A, in this state it won't receive more damage and won't deal damage or won't collide, it is like `passive` we've discussed.
- Invulnerable State B: If a player collects a power-up or something, it just won't receive any damage, but it still deals damage to the others.

Is there any possibility we find two words to well describe them?

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:671]: > > I am sometimes too obsessive
> 
> I definitely can't blame you for _caring_ about the future of the game! I confirm I was too harsh with my initial comment, and I'd like to apologise for this.
> 
> These tests are indeed a bit premature at this stage, but they don't remove anything or add anything dangerous to the game AFAICS and may pave the way for the next steps with the passive state. Take the above as advice on how to better design tests, but what you're doing is an important step too.

This mechanism really involves a lot of worth-thinking issues, I've been a bit busy with my paper recently, I would find a time to rethink about it, and try to improve this PR

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:738]: > I think you've figured out my opinion already 😉 I believe anticipating possible extensions beyond the minimal viable game is premature optimisation, and a wise man once told me it's the [root of all evil](https://effectiviology.com/premature-optimization/).
> 
> So for now, I'd rather have the duration as a fixed parameter. It will be easy to make it free again when/if we need it :)

I'll change it in the following commits

File: src/game.cpp
Comment (TheoPannetier) [reply src/game.cpp:147]: > Invulnerable State B: If a player collects a power-up or something, it just won't receive any damage, but it still deals damage to the others

We don't have power-ups are plans to add any before `1.0`, so let's not worry about it for now ;)
We can always make changes later if the current implementation becomes a barrier to new desired features.
Let's keep it to what's referred to in the title of this PR.
So for this "invulnerable state A", I guess either `passive` or `invulnerable` are ok names; neither is entirely satisfying but they do an ok job.

**Comment by TheoPannetier** [comment]
Hello @EvoLandEco !
I think the discussion has been a bit dense and dispersed, so to make your life easier here are the two things I'd like to see addressed before I accept this PR:

1. When resolving collision and player growth/shrinkage, you check that players involved in the collision are both active before resolving the effects of the collision. This is good, but you do the check inside the growing/shrinking functions. I feel that if either player is not active, there should not be any collision at all (and no need to call the growth/shrinkage functions). So the check for player active/passive state should be done earlier, and `has_any_player_collision` should return FALSE if either player is passive.
2. Make passive duration a fixed parameter of either the `player` or `game`, and remove the corresponding setter.

(and as usual, resolve any pending merge conflict!)

File: src/game.cpp
Comment (EvoLandEco) [reply src/game.cpp:147]: @TheoPannetier That's true, I will refactor `passive` back to `invulnerable` for now

**Comment by EvoLandEco** [comment]
> Hello @EvoLandEco ! I think the discussion has been a bit dense and dispersed, so to make your life easier here are the two things I'd like to see addressed before I accept this PR:
> 
> 1. When resolving collision and player growth/shrinkage, you check that players involved in the collision are both active before resolving the effects of the collision. This is good, but you do the check inside the growing/shrinking functions. I feel that if either player is not active, there should not be any collision at all (and no need to call the growth/shrinkage functions). So the check for player active/passive state should be done earlier, and `has_any_player_collision` should return FALSE if either player is passive.
> 2. Make passive duration a fixed parameter of either the `player` or `game`, and remove the corresponding setter.
> 
> (and as usual, resolve any pending merge conflict!)

Let's focus on what this PR is meant to be. 😉 I remember I did status check in a later stage due to a limitation from how collision check was implemented, but I will surely revisit this part and see if I can do any better.

For the passive duration, I agree to make it align with what we did in the shoot cooldown system: maybe make it a `static` variable.

**Comment by EvoLandEco** [comment]
> Hello @EvoLandEco ! I think the discussion has been a bit dense and dispersed, so to make your life easier here are the two things I'd like to see addressed before I accept this PR:
> 
> 1. When resolving collision and player growth/shrinkage, you check that players involved in the collision are both active before resolving the effects of the collision. This is good, but you do the check inside the growing/shrinking functions. I feel that if either player is not active, there should not be any collision at all (and no need to call the growth/shrinkage functions). So the check for player active/passive state should be done earlier, and `has_any_player_collision` should return FALSE if either player is passive.
> 2. Make passive duration a fixed parameter of either the `player` or `game`, and remove the corresponding setter.
> 
> (and as usual, resolve any pending merge conflict!)

@TheoPannetier We have the `game_options` class here, what about putting passive duration in this class?

**Comment by TheoPannetier** [comment]
> @TheoPannetier We have the game_options class here, what about putting passive duration in this class?

Indeed, it looks appropriate :)
(sorry, only seeing this now)

**Comment by TheoPannetier** [comment]
Merged to not lose progress

**Pull request merged**
